### PR TITLE
Use enums instead of strings to denote memory space

### PR DESCRIPTION
**Pull request by pelesh**
Replace strings that are currently used to identify different memory spaces with `enum`s. 

`enum`s are now used to select memory space for vector and matrix classes. The options are HOST and DEVICE. 

This PR does not deal with matrix and vector _handlers_.

### CODE REVIEW

**Review by cameronrutherford — APPROVED** [review]

**Review by rothpc — APPROVED** [review]

**Comment by pelesh** [comment]
After these changes unit tests will need some refactoring. They are patched up and functional now, but not quite straightforward to maintan.

**Comment by pelesh** [comment]
Related issue is #12

**Review by kswirydo — APPROVED** [review]
1) Compiles and runs on ascent
2) Minor comments added

File: resolve/matrix/Coo.cpp [hunk 1]
```diff
@@ -27,65 +27,65 @@ namespace ReSolve
...
-      return this->h_row_data_;
-    } else {
-      if ((memspace == "cuda") || (memspace == "hip")) {
+    using namespace ReSolve::memory;
+    switch (memspace) {
+      case HOST:
+        copyData(memspace);
+        return this->h_row_data_;
+      case DEVICE:
         copyData(memspace);
```
[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/matrix/Coo.cpp]: Fixed in [af16894](https://github.com/ORNL/ReSolve/pull/42/commits/af1689407be7c3ec8886ac1cfd465a3dbc0c1111)

**Pull request merged**